### ISSUE TITLE
Differentiate through the rendering process

### ISSUE THREAD

**Issue by Wuziyi616** [kwea123/ngp_pl#9]
Hi, thanks for the great work! I have a (maybe stupid) question regarding the implementation. I'm not a NeRF researcher, but there are many works in my field adopting the NeRF representation for different applications. When using NeRF, we mostly care about the gradients from rendered RGB values (or depth) w.r.t. to input rays (or query points etc.). I looked at the official instant-ngp repo before, and I find out we cannot easily differentiate through their CUDA code.

Since this is a PyTorch-based implementation, I'm wondering if we can easily get the gradient. I haven't thoroughly read the code, but as far as I read the answer seems to be yes. After all you will need PyTorch auto-grad mechanism to "train" the network. Is that correct?

**Comment by kwea123** [comment]
> I find out we cannot easily differentiate ...

I don't understand what you mean by "easily differentiate", you mean no autograd? Yes, in cuda you need to first derive the gradient mathematically and implement that. That's what instant-ngp does [here](https://github.com/NVlabs/instant-ngp/blob/master/src/testbed_nerf.cu#L1477-L1519)
And in my implementation there is no "easy" way either, I have to derive the formula, then implement:
https://github.com/kwea123/ngp_pl/blob/e9d9c37fa0251104332fa0950f80221d1cfea203/models/csrc/volumerendering.cu#L137-L152

**Comment by Wuziyi616** [comment]
Thanks for the reply. Yes, I understand the CUDA implementation things. What I'm trying to say is that, since this codebase is based on PyTorch, that means you have wrapped the CUDA kernels using Python+PyTorch (correct?). So if I want to integrate it into my codebase (I'll treat the NeRF part as a black-box, basically calculating the gradients over e.g. RGB MSE loss), it would be easier than the official implementation.

Sorry for the confusion in my previous post.
